From 4557aae134be0463d2634b672cd9e6355ff1ad4c Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zdoychev@maxlinear.com>
Date: Mon, 15 Mar 2021 17:14:44 +0100
Subject: [PATCH] dp: pon_qos: add trace events

PONRTSYS-8930: Add trace events to the pon qos driver. This allows
debugging the configuration flows without much impact on system
performance and timings.
---
 drivers/net/datapath/pon_qos/Kconfig               |   7 +
 drivers/net/datapath/pon_qos/Makefile              |   5 +
 drivers/net/datapath/pon_qos/pon_qos_tc_drr.c      |   4 +
 drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c |  18 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_flower.c   |   5 +
 drivers/net/datapath/pon_qos/pon_qos_tc_main.c     |  35 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_prio.c     |   6 +
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c      |  27 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_red.c      |  13 +-
 drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c      |   5 +
 .../net/datapath/pon_qos/pon_qos_tc_vlan_filter.c  |  18 +-
 drivers/net/datapath/pon_qos/pon_qos_trace.c       |  11 +
 drivers/net/datapath/pon_qos/pon_qos_trace.h       | 637 +++++++++++++++++++++
 13 files changed, 774 insertions(+), 17 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/Kconfig b/drivers/net/datapath/pon_qos/Kconfig
index 87649b245c7d..5247ab616b2c 100644
--- a/drivers/net/datapath/pon_qos/Kconfig
+++ b/drivers/net/datapath/pon_qos/Kconfig
@@ -17,3 +17,10 @@ config PON_QOS_KCOV
 	depends on KCOV && PON_QOS
 	help
 	  Enable coverage on PON QoS driver for fuzzing.
+
+config PON_QOS_TRACING
+	bool "PON QoS tracing support"
+	depends on PON_QOS
+	depends on EVENT_TRACING
+	---help---
+	  Select this to enable pon_qos tracing infrastructure.
diff --git a/drivers/net/datapath/pon_qos/Makefile b/drivers/net/datapath/pon_qos/Makefile
index 59560e52a162..bedb88b533d8 100644
--- a/drivers/net/datapath/pon_qos/Makefile
+++ b/drivers/net/datapath/pon_qos/Makefile
@@ -24,3 +24,8 @@ pon_qos-objs := \
 	pon_qos_tc_tbf.o \
 	pon_qos_tc_trap.o \
 	pon_qos_tc_ip_drop.o
+
+pon_qos-$(CONFIG_PON_QOS_TRACING) += pon_qos_trace.o
+
+# otherwise trace header is not found
+CFLAGS_pon_qos_trace.o += -I$(src)
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c b/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c
index e5f667fa2852..6dd8335870fb 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_drr.c
@@ -9,6 +9,7 @@
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
 #include "pon_qos_tc_qos.h"
+#include "pon_qos_trace.h"
 
 #define MAX_QUANTUM (255 * 4096)
 #define KB 1024
@@ -295,6 +296,8 @@ int pon_qos_tc_drr_offload(struct net_device *dev,
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "DRR: offload starting\n");
+
+	trace_pon_qos_tc_drr_enter(dev, opt);
 	switch (opt->command) {
 	case TC_DRR_REPLACE:
 		netdev_dbg(dev, "pid: %#x class/handle: %#x quantum: %#x\n",
@@ -326,6 +329,7 @@ int pon_qos_tc_drr_offload(struct net_device *dev,
 	default:
 		break;
 	}
+	trace_pon_qos_tc_drr_exit(dev, opt);
 
 	netdev_dbg(dev, "DRR: offload end: %d\n", err);
 	return err;
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
index 8dc82f8961e2..26eec69b36ff 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
@@ -20,6 +20,7 @@
 #include "pon_qos_tc_vlan_storage.h"
 #include "pon_qos_tc_ext_vlan.h"
 #include "pon_qos_tc_lct.h"
+#include "pon_qos_trace.h"
 
 static LIST_HEAD(tc_ext_vlan_storage);
 
@@ -368,6 +369,8 @@ int pon_qos_tc_ext_vlan_add(struct net_device *dev,
 	if (tag == TC_VLAN_UNKNOWN)
 		return -EIO;
 
+	trace_ev_add_enter(dev, NULL, -1, -1, ingress);
+
 	node = pon_qos_tc_vlan_storage_get(dev, ingress, tag,
 					   &tc_ext_vlan_storage);
 	if (!node) {
@@ -403,7 +406,11 @@ int pon_qos_tc_ext_vlan_add(struct net_device *dev,
 	netdev_dbg(dev, "Extended VLAN rule added, storage: %p, ingress: %d, tag: %d, rule_cnt: %d\n",
 		   node, node->ingress, node->tag, node->rule_cnt);
 
-	return dp_update(dev, &dp_vlan, ingress);
+	err = dp_update(dev, &dp_vlan, ingress);
+
+	trace_ev_add_exit(dev, node, node->tag, node->rule_cnt, ingress);
+
+	return err;
 err:
 	pon_qos_tc_vlan_storage_del(dev, node);
 	return err;
@@ -419,6 +426,7 @@ int pon_qos_tc_ext_vlan_del(struct net_device *dev,
 	struct dp_vlan1 *rule1;
 	struct dp_vlan2 *rule2;
 	bool ingress;
+	int ret = 0;
 
 	netdev_dbg(dev, "Extended VLAN delete start, node: %p, rule: %p\n",
 		   vlan_storage, rule);
@@ -429,6 +437,8 @@ int pon_qos_tc_ext_vlan_del(struct net_device *dev,
 	node = vlan_storage;
 	ingress = node->ingress;
 
+	trace_ev_del_enter(dev, node, node->tag, node->rule_cnt, ingress);
+
 	pon_qos_tc_dp_vlan_prepare(dev, &dp_vlan, ingress);
 
 	switch (node->tag) {
@@ -457,7 +467,11 @@ int pon_qos_tc_ext_vlan_del(struct net_device *dev,
 
 	pon_qos_tc_vlan_storage_del(dev, node);
 
-	return dp_update(dev, &dp_vlan, ingress);
+	ret = dp_update(dev, &dp_vlan, ingress);
+
+	trace_ev_del_exit(dev, node, node->tag, node->rule_cnt, ingress);
+
+	return ret;
 }
 
 struct list_head *pon_qos_tc_get_ext_vlan_storage(void)
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c b/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c
index 6b17b045c063..c4e95954550f 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_flower.c
@@ -19,6 +19,7 @@
 #include "pon_qos_tc_police.h"
 #include "pon_qos_tc_trap.h"
 #include "pon_qos_tc_ip_drop.h"
+#include "pon_qos_trace.h"
 
 struct pon_qos_storage_node {
 	/** Unique key for TC flower - flower cookie */
@@ -336,18 +337,21 @@ int pon_qos_tc_flower_offload(struct net_device *dev,
 	case TC_CLSFLOWER_REPLACE:
 	{
 		netdev_dbg(dev, "%s:replace:\n", __func__);
+		trace_pon_qos_tc_flower_enter(dev, f, flower_type_get(dev, f));
 		ret = pon_qos_tc_flower_replace(dev, handle, f);
 		break;
 	}
 	case TC_CLSFLOWER_DESTROY:
 	{
 		netdev_dbg(dev, "%s:destroy:\n", __func__);
+		trace_pon_qos_tc_flower_enter(dev, f, -1);
 		ret = pon_qos_tc_flower_destroy(dev, handle, f);
 		break;
 	}
 	case TC_CLSFLOWER_STATS:
 	{
 		netdev_dbg(dev, "%s:stats:\n", __func__);
+		trace_pon_qos_tc_flower_enter(dev, f, -1);
 		return -EOPNOTSUPP;
 	}
 	default:
@@ -355,5 +359,6 @@ int pon_qos_tc_flower_offload(struct net_device *dev,
 	}
 
 	netdev_dbg(dev, "%s:exit %d\n", __func__, ret);
+	trace_pon_qos_tc_flower_exit(dev, f, -1);
 	return ret;
 }
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_main.c b/drivers/net/datapath/pon_qos/pon_qos_tc_main.c
index a0fecf2b6e46..4c668af38ac4 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_main.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_main.c
@@ -7,9 +7,11 @@
  *****************************************************************************/
 #include <net/pkt_cls.h>
 #include <net/pon_qos_tc.h>
+#include <linux/jiffies.h>
 #include "pon_qos_tc_main.h"
 #include "pon_qos_tc_qos.h"
 #include "pon_qos_tc_debugfs.h"
+#include "pon_qos_trace.h"
 #if IS_ENABLED(CONFIG_QOS_MGR)
 #include <net/qos_mgr/qos_mgr_tc_hook.h>
 #endif
@@ -27,30 +29,41 @@ int pon_qos_setup_tc(struct net_device *dev,
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "%s: start\n", __func__);
+
+	trace_pon_qos_setup_tc_enter(dev, tc->type);
+
 	switch (tc->type) {
 	case TC_SETUP_CLSFLOWER:
-		return pon_qos_tc_flower_offload(dev,
-						 handle,
-						 tc->cls_flower);
+		ret = pon_qos_tc_flower_offload(dev,
+						handle,
+						tc->cls_flower);
+		break;
 	case TC_SETUP_MQPRIO:
-		return pon_qos_tc_mqprio_offload(dev, handle, tc,
-						 port_id, deq_idx);
+		ret = pon_qos_tc_mqprio_offload(dev, handle, tc,
+						port_id, deq_idx);
+		break;
 	case TC_SETUP_QDISC_PRIO:
-		return pon_qos_tc_prio_offload(dev, handle, tc,
-					       port_id, deq_idx);
-	case TC_SETUP_DRR:
-		return pon_qos_tc_drr_offload(dev, handle, tc,
+		ret = pon_qos_tc_prio_offload(dev, handle, tc,
 					      port_id, deq_idx);
+		break;
+	case TC_SETUP_DRR:
+		ret = pon_qos_tc_drr_offload(dev, handle, tc,
+					     port_id, deq_idx);
+		break;
 	case TC_SETUP_QDISC_RED:
-		return pon_qos_tc_red_offload(dev, handle, tc);
+		ret = pon_qos_tc_red_offload(dev, handle, tc);
+		break;
 	case TC_SETUP_TBF:
-		return pon_qos_tc_tbf_offload(dev, handle, tc);
+		ret = pon_qos_tc_tbf_offload(dev, handle, tc);
+		break;
 	default:
 		netdev_err(dev, "offload type:%d not supported\n", tc->type);
 		ret = -EOPNOTSUPP;
 		break;
 	}
 
+	trace_pon_qos_setup_tc_exit(dev, tc->type);
+
 	return ret;
 }
 EXPORT_SYMBOL(pon_qos_setup_tc);
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c b/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c
index 06010adc1ccb..a8a5c35d6e73 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_prio.c
@@ -9,6 +9,7 @@
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
 #include "pon_qos_tc_qos.h"
+#include "pon_qos_trace.h"
 
 /* create qdisc tree */
 
@@ -181,6 +182,9 @@ int pon_qos_tc_prio_offload(struct net_device *dev,
 
 	ASSERT_RTNL();
 	netdev_dbg(dev, "PRIO: offload starting\n");
+
+	trace_pon_qos_tc_prio_enter(dev, opt);
+
 	switch (opt->command) {
 	case TC_PRIO_REPLACE:
 		if (opt->replace_params.bands > PON_QOS_TC_MAX_Q) {
@@ -220,6 +224,8 @@ int pon_qos_tc_prio_offload(struct net_device *dev,
 		break;
 	}
 
+	trace_pon_qos_tc_prio_exit(dev, opt);
+
 	netdev_dbg(dev, "PRIO: offload end: %d\n", err);
 	return 0;
 }
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index c20f46170e05..1707e9f0a33f 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -11,6 +11,7 @@
 #include <net/pon_qos_tc.h>
 #include "pon_qos_tc_qos.h"
 #include "pon_qos_tc_qmap.h"
+#include "pon_qos_trace.h"
 
 static LIST_HEAD(port_list);
 
@@ -248,6 +249,8 @@ int pon_qos_add_sched(struct pon_qos_qdisc *sch, int prio)
 	if (sch->parent != TC_H_ROOT)
 		return -EINVAL;
 
+	trace_pon_qos_add_sched_exit(sch, 0);
+
 	/* change def queue prio */
 	prio_info.inst = sch->inst;
 	prio_info.id.q_id = sch->def_q;
@@ -305,6 +308,8 @@ int pon_qos_add_sched(struct pon_qos_qdisc *sch, int prio)
 		}
 	}
 
+	trace_pon_qos_add_sched_exit(sch, 0);
+
 	return 0;
 }
 
@@ -316,6 +321,8 @@ int pon_qos_add_staged_sched(struct pon_qos_qdisc *psch,
 	struct dp_node_alloc anode = {0};
 	int ret;
 
+	trace_pon_qos_add_sched_enter(csch, psch->sch_id);
+
 	/* Allocate sched */
 	anode.inst = csch->inst;
 	anode.dp_port = csch->port;
@@ -349,6 +356,8 @@ int pon_qos_add_staged_sched(struct pon_qos_qdisc *psch,
 		return -ENODEV;
 	}
 
+	trace_pon_qos_add_sched_exit(csch, psch->sch_id);
+
 	return 0;
 }
 
@@ -358,6 +367,8 @@ int pon_qos_sched_del(struct pon_qos_qdisc *sch)
 	struct dp_node_alloc anode = {0};
 	int ret;
 
+	trace_pon_qos_sched_del_enter(sch, 0);
+
 	node.node_type = DP_NODE_SCH;
 	node.node_id.sch_id = sch->sch_id;
 	ret = dp_node_unlink(&node, 0);
@@ -378,6 +389,8 @@ int pon_qos_sched_del(struct pon_qos_qdisc *sch)
 
 	sch->use_cnt = 0;
 
+	trace_pon_qos_sched_del_exit(sch, 0);
+
 	return 0;
 }
 
@@ -605,6 +618,8 @@ int pon_qos_queue_add(struct pon_qos_qdisc *sch, int arbi, int prio_w, int idx)
 	if (!sch)
 		return -EINVAL;
 
+	trace_pon_qos_queue_add_enter(sch, &sch->qids[idx], idx);
+
 	anode.inst = sch->inst;
 	anode.dp_port = sch->port;
 	anode.type = DP_NODE_QUEUE;
@@ -648,6 +663,8 @@ int pon_qos_queue_add(struct pon_qos_qdisc *sch, int arbi, int prio_w, int idx)
 	if (ret < 0)
 		netdev_err(sch->dev, "%s:wred defaults set fail\n", __func__);
 
+	trace_pon_qos_queue_add_exit(sch, &sch->qids[idx], idx);
+
 	return 0;
 }
 
@@ -666,6 +683,8 @@ int pon_qos_queue_del(struct pon_qos_qdisc *sch, int idx)
 	if (idx < 0 || idx > PON_QOS_TC_MAX_Q - 1)
 		return -EINVAL;
 
+	trace_pon_qos_queue_del_enter(sch, &sch->qids[idx], idx);
+
 	ret = pon_qos_qdata_child_remove(sch->dev, &sch->qids[idx],
 					 sch->handle);
 	if (ret)
@@ -681,6 +700,8 @@ int pon_qos_queue_del(struct pon_qos_qdisc *sch, int idx)
 		netdev_err(sch->dev, "qid %d free failed\n",
 			   anode.id.q_id);
 
+	trace_pon_qos_queue_del_exit(sch, &sch->qids[idx], idx);
+
 	memset(&sch->qids[idx], 0, sizeof(struct pon_qos_q_data));
 	sch->num_q--;
 
@@ -997,6 +1018,8 @@ int pon_qos_update_qmap(struct net_device *dev,
 		   q_tc->tc, q_tc->tc_cookie,
 		   sch->ds ? "true" : "false", en);
 
+	trace_pon_qos_update_qmap_enter(dev, sch, qid, q_tc, en);
+
 	/* TODO: DPM uses port zero currently as the CPU port */
 	if (pon_qos_is_cpu_port(sch->port))
 		return pon_qos_config_cpu_port(qid, sch->port, 0,
@@ -1010,7 +1033,9 @@ int pon_qos_update_qmap(struct net_device *dev,
 	}
 
 	if (sch->ds)
-		return pon_qos_update_ds_qmap(dev, sch, q_tc, en);
+		ret = pon_qos_update_ds_qmap(dev, sch, q_tc, en);
+
+	trace_pon_qos_update_qmap_exit(dev, sch, qid, q_tc, en);
 
 	return ret;
 }
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_red.c b/drivers/net/datapath/pon_qos/pon_qos_tc_red.c
index 9711576cadce..346559a07f6b 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_red.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_red.c
@@ -10,6 +10,7 @@
 #include <net/datapath_api.h>
 #include <net/datapath_api_qos.h>
 #include "pon_qos_tc_qos.h"
+#include "pon_qos_trace.h"
 
 #define WRED_GREEN_MAX_TH (500 * 1024)
 #define WRED_GREEN_MIN_TH ((WRED_GREEN_MAX_TH * 90) / 100)
@@ -198,6 +199,8 @@ static int pon_qos_tc_red_replace(struct net_device *dev,
 		return ret;
 	}
 
+	trace_pon_qos_tc_red_replace(dev, opt, qdisc, qid, type);
+
 	return 0;
 }
 
@@ -234,7 +237,11 @@ static int pon_qos_tc_red_destroy(struct net_device *dev,
 		}
 	}
 
-	return pon_qos_qdata_remove(dev, qid, handle, opt->parent);
+	ret = pon_qos_qdata_remove(dev, qid, handle, opt->parent);
+
+	trace_pon_qos_tc_red_destroy(dev, opt, qdisc, qid);
+
+	return ret;
 }
 
 int pon_qos_tc_red_offload(struct net_device *dev,
@@ -247,6 +254,8 @@ int pon_qos_tc_red_offload(struct net_device *dev,
 	ASSERT_RTNL();
 	netdev_dbg(dev, "RED: offload starting\n");
 
+	trace_pon_qos_tc_red_enter(dev, opt);
+
 	switch (opt->command) {
 	case TC_RED_REPLACE:
 		netdev_dbg(dev, "replace pid:%#x class/handle:%#x\n",
@@ -273,6 +282,8 @@ int pon_qos_tc_red_offload(struct net_device *dev,
 		break;
 	}
 
+	trace_pon_qos_tc_red_exit(dev, opt);
+
 	netdev_dbg(dev, "RED: offload end: %d\n", err);
 	return 0;
 }
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c b/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c
index 7ae2bd5c5b8e..cd86d9fb39d5 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_tbf.c
@@ -6,6 +6,7 @@
  *****************************************************************************/
 #include <net/pkt_cls.h>
 #include "pon_qos_tc_qos.h"
+#include "pon_qos_trace.h"
 
 static int pon_tbf_remove(struct net_device *dev, u32 handle, u32 parent)
 {
@@ -138,6 +139,8 @@ int pon_qos_tc_tbf_offload(struct net_device *dev,
 	int err = 0;
 	struct tc_tbf_qopt_offload *opt = tc->sch_tbf;
 
+	trace_pon_qos_tc_tbf_enter(dev, opt);
+
 	switch (opt->command) {
 	case TC_TBF_REPLACE:
 		err = pon_qos_tc_tbf_replace(dev, opt);
@@ -160,5 +163,7 @@ int pon_qos_tc_tbf_offload(struct net_device *dev,
 		break;
 	}
 
+	trace_pon_qos_tc_tbf_exit(dev, opt);
+
 	return 0;
 }
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c
index 2e8adb361519..c279dfaf1092 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_vlan_filter.c
@@ -22,6 +22,7 @@
 #include "pon_qos_tc_vlan_filter.h"
 #include "pon_qos_tc_lct.h"
 #include "pon_qos_tc_ext_vlan.h"
+#include "pon_qos_trace.h"
 
 #define PON_QOS_TC_FILTER_DEF_PRIO_THRESHOLD 64000
 
@@ -187,6 +188,8 @@ int pon_qos_tc_vlan_filter_add(struct net_device *dev,
 
 	netdev_dbg(dev, "VLAN filter add, flower: %p\n", f);
 
+	trace_vf_add_enter(dev, NULL, -1, -1, ingress);
+
 	tag = pon_qos_tc_vlan_tag_get(dev, f);
 	if (tag == TC_VLAN_UNKNOWN)
 		return -EIO;
@@ -223,7 +226,11 @@ int pon_qos_tc_vlan_filter_add(struct net_device *dev,
 		goto err;
 	}
 
-	return dp_update(dev, &dp_vlan, ingress);
+	err = dp_update(dev, &dp_vlan, ingress);
+
+	trace_vf_add_exit(dev, node, node->tag, node->rule_cnt, ingress);
+
+	return err;
 err:
 	pon_qos_tc_vlan_storage_del(dev, node);
 	return err;
@@ -239,6 +246,7 @@ int pon_qos_tc_vlan_filter_del(struct net_device *dev,
 	struct dp_vlan1 *rule1;
 	struct dp_vlan2 *rule2;
 	bool ingress;
+	int ret = 0;
 
 	netdev_dbg(dev, "VLAN filter del, node: %p, rule: %p\n",
 		   vlan_storage, rule);
@@ -249,6 +257,8 @@ int pon_qos_tc_vlan_filter_del(struct net_device *dev,
 	node = vlan_storage;
 	ingress = node->ingress;
 
+	trace_vf_del_enter(dev, node, node->tag, node->rule_cnt, ingress);
+
 	pon_qos_tc_dp_vlan_prepare(dev, &dp_vlan, ingress);
 
 	switch (node->tag) {
@@ -273,7 +283,11 @@ int pon_qos_tc_vlan_filter_del(struct net_device *dev,
 	node->rule_cnt--;
 	pon_qos_tc_vlan_storage_del(dev, node);
 
-	return dp_update(dev, &dp_vlan, ingress);
+	ret = dp_update(dev, &dp_vlan, ingress);
+
+	trace_vf_del_exit(dev, NULL, -1, -1, ingress);
+
+	return ret;
 }
 
 static bool pon_qos_tc_lookup_by_cvid(struct pon_qos_vlan_storage_node *node,
diff --git a/drivers/net/datapath/pon_qos/pon_qos_trace.c b/drivers/net/datapath/pon_qos/pon_qos_trace.c
new file mode 100644
index 000000000000..83e6857896b7
--- /dev/null
+++ b/drivers/net/datapath/pon_qos/pon_qos_trace.c
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: GPL-2.0
+/******************************************************************************
+ *
+ * Copyright (c) 2021 MaxLinear, Inc.
+ * Copyright (c) 2020 Intel Corporation
+ *
+ *****************************************************************************/
+#include <linux/module.h>
+
+#define CREATE_TRACE_POINTS
+#include "pon_qos_trace.h"
diff --git a/drivers/net/datapath/pon_qos/pon_qos_trace.h b/drivers/net/datapath/pon_qos/pon_qos_trace.h
new file mode 100644
index 000000000000..d580da640a47
--- /dev/null
+++ b/drivers/net/datapath/pon_qos/pon_qos_trace.h
@@ -0,0 +1,637 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/******************************************************************************
+ *
+ * Copyright (c) 2021 MaxLinear, Inc.
+ *
+ *****************************************************************************/
+#undef TRACE_SYSTEM
+#define TRACE_SYSTEM pon_qos_tc
+
+#if !defined(_PON_QOS_TRACE_H_) || defined(TRACE_HEADER_MULTI_READ)
+#define _PON_QOS_TRACE_H_
+
+#include <linux/tracepoint.h>
+#include <net/pkt_cls.h>
+
+#include "pon_qos_tc_qos.h"
+#include "pon_qos_tc_flower.h"
+
+/* create empty functions when tracing is disabled */
+#if !defined(CONFIG_PON_QOS_TRACING)
+#undef TRACE_EVENT
+#define TRACE_EVENT(name, proto, ...) \
+static inline void trace_ ## name(proto) {}
+#undef DECLARE_EVENT_CLASS
+#define DECLARE_EVENT_CLASS(...)
+#undef DEFINE_EVENT
+#define DEFINE_EVENT(evt_class, name, proto, ...) \
+static inline void trace_ ## name(proto) {}
+#endif
+
+DECLARE_EVENT_CLASS(pon_qos_tc_class,
+	TP_PROTO(const struct net_device *dev, const unsigned int type),
+
+	TP_ARGS(dev, type),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(unsigned int, type)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->type = type;
+	),
+
+	TP_printk("[%s] -> type: %u", __get_str(name), __entry->type)
+);
+
+DEFINE_EVENT(pon_qos_tc_class, pon_qos_setup_tc_enter,
+	TP_PROTO(const struct net_device *dev, const unsigned int type),
+	TP_ARGS(dev, type)
+);
+
+DEFINE_EVENT(pon_qos_tc_class, pon_qos_setup_tc_exit,
+	TP_PROTO(const struct net_device *dev, const unsigned int type),
+	TP_ARGS(dev, type)
+);
+
+#define pon_qos_tc_offload_type				\
+	{ TC_TYPE_EXT_VLAN, "TC_TYPE_EXT_VLAN" },	\
+	{ TC_TYPE_VLAN_FILTER, "TC_TYPE_VLAN_FILTER" }, \
+	{ TC_TYPE_QUEUE, "TC_TYPE_QUEUE" },		\
+	{ TC_TYPE_MIRRED, "TC_TYPE_MIRRED" },		\
+	{ TC_TYPE_POLICE, "TC_TYPE_POLICE" },		\
+	{ TC_TYPE_COLMARK, "TC_TYPE_COLMARK" },		\
+	{ TC_TYPE_TRAP, "TC_TYPE_TRAP" },		\
+	{ TC_TYPE_IP_DROP, "TC_TYPE_DROP" },		\
+	{ -EOPNOTSUPP, "TC_TYPE_UNKNOWN"  }
+
+DECLARE_EVENT_CLASS(pon_qos_tc_flower_class,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_cls_flower_offload *f,
+		 int type),
+
+	TP_ARGS(dev, f, type),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u8, command)
+		__field(unsigned long, cookie)
+		__field(unsigned int, classid)
+		__field(int, type)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->command = f->command;
+		__entry->cookie = f->cookie;
+		__entry->classid = f->classid;
+		__entry->type = type;
+	),
+
+	TP_printk("[%s] -> cmd: %#x cookie: %#lx classid: %#x type: %s",
+		  __get_str(name), __entry->command,
+		  __entry->cookie, __entry->classid,
+		  __print_symbolic(__entry->type, pon_qos_tc_offload_type))
+);
+
+DEFINE_EVENT(pon_qos_tc_flower_class, pon_qos_tc_flower_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_cls_flower_offload *f,
+		 int type),
+	TP_ARGS(dev, f, type)
+);
+
+DEFINE_EVENT(pon_qos_tc_flower_class, pon_qos_tc_flower_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_cls_flower_offload *f,
+		 int type),
+	TP_ARGS(dev, f, type)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_tc_drr_class,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_drr_qopt_offload *opt),
+
+	TP_ARGS(dev, opt),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u8, command)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(int, quantum)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->command = opt->command;
+		__entry->handle = opt->handle;
+		__entry->parent = opt->parent;
+		__entry->quantum = (opt->command == TC_DRR_STATS) ?
+				    0 : opt->set_params.quantum;
+	),
+
+	TP_printk("[%s] -> cmd: %#x parent: %#x handle: %#x quantum: %d",
+		  __get_str(name), __entry->command,
+		  __entry->parent, __entry->handle,
+		  __entry->quantum)
+);
+
+DEFINE_EVENT(pon_qos_tc_drr_class, pon_qos_tc_drr_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_drr_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DEFINE_EVENT(pon_qos_tc_drr_class, pon_qos_tc_drr_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_drr_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_tc_prio_class,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_prio_qopt_offload *opt),
+
+	TP_ARGS(dev, opt),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u8, command)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(int, bands)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->command = opt->command;
+		__entry->handle = opt->handle;
+		__entry->parent = opt->parent;
+		__entry->bands = (opt->command == TC_PRIO_REPLACE) ?
+				  opt->replace_params.bands :
+				 (opt->command != TC_PRIO_GRAFT) ?
+				  0 : opt->graft_params.band;
+	),
+
+	TP_printk("[%s] -> cmd:%#x parent:%#x handle:%#x band(s):%#x",
+		  __get_str(name), __entry->command,
+		  __entry->parent, __entry->handle,
+		  __entry->bands)
+);
+
+DEFINE_EVENT(pon_qos_tc_prio_class, pon_qos_tc_prio_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_prio_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DEFINE_EVENT(pon_qos_tc_prio_class, pon_qos_tc_prio_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_prio_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_tc_tbf_class,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_tbf_qopt_offload *opt),
+
+	TP_ARGS(dev, opt),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u8, command)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(u64, rate)
+		__field(u64, prate)
+		__field(u32, burst)
+		__field(u32, pburst)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->command = opt->command;
+		__entry->handle = opt->handle;
+		__entry->parent = opt->parent;
+		__entry->rate = (opt->command == TC_TBF_REPLACE) ?
+				 opt->set_params.rate : 0;
+		__entry->prate = (opt->command == TC_TBF_REPLACE) ?
+				     opt->set_params.prate : 0;
+		__entry->burst = (opt->command == TC_TBF_REPLACE) ?
+				  opt->set_params.burst : 0;
+		__entry->pburst = (opt->command == TC_TBF_REPLACE) ?
+				   opt->set_params.pburst : 0;
+	),
+
+	TP_printk("[%s] -> cmd:%#x parent:%#x handle:%#x rate:%llu peakrate:%llu burst:%u pburst:%u",
+		  __get_str(name), __entry->command,
+		  __entry->parent, __entry->handle,
+		  __entry->rate, __entry->prate,
+		  __entry->burst, __entry->pburst)
+);
+
+DEFINE_EVENT(pon_qos_tc_tbf_class, pon_qos_tc_tbf_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_tbf_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DEFINE_EVENT(pon_qos_tc_tbf_class, pon_qos_tc_tbf_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_tbf_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_tc_red_class,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_red_qopt_offload *opt),
+
+	TP_ARGS(dev, opt),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u8, command)
+		__field(u32, handle)
+		__field(u32, parent)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->command = opt->command;
+		__entry->handle = opt->handle;
+		__entry->parent = opt->parent;
+	),
+
+	TP_printk("[%s] -> cmd:%#x parent:%#x handle:%#x",
+		  __get_str(name), __entry->command,
+		  __entry->parent, __entry->handle)
+);
+
+DEFINE_EVENT(pon_qos_tc_red_class, pon_qos_tc_red_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_red_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DEFINE_EVENT(pon_qos_tc_red_class, pon_qos_tc_red_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_red_qopt_offload *opt),
+	TP_ARGS(dev, opt)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_sched_class,
+	TP_PROTO(const struct pon_qos_qdisc *sch, u32 pid),
+
+	TP_ARGS(sch, pid),
+
+	TP_STRUCT__entry(
+		__string(name, sch->dev->name)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(u32, sch_id)
+		__field(int, p_w)
+		__field(s32, port)
+		__field(int, epn)
+		__field(u16, def_q)
+		__field(int, use_cnt)
+		__field(unsigned int, num_q)
+		__field(unsigned int, num_children)
+		__field(u32, pid)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, sch->dev->name);
+		__entry->handle = sch->handle;
+		__entry->parent = sch->handle;
+		__entry->sch_id = sch->sch_id;
+		__entry->p_w = sch->p_w;
+		__entry->port = sch->port;
+		__entry->epn = sch->epn;
+		__entry->def_q = sch->def_q;
+		__entry->use_cnt = sch->use_cnt;
+		__entry->num_q = sch->num_q;
+		__entry->num_children = sch->num_children;
+		__entry->pid = pid;
+	),
+
+	TP_printk("[%s] hdls:%#x/%#x id:%u p_w:%d port:%i epn:%i def_q:%u psch_id: %u use_cnt: %d num_q: %u children: %u",
+		  __get_str(name), __entry->parent, __entry->handle,
+		  __entry->sch_id, __entry->p_w, __entry->port,
+		  __entry->epn, __entry->def_q, __entry->pid,
+		  __entry->use_cnt, __entry->num_q, __entry->num_children)
+);
+
+DEFINE_EVENT(pon_qos_sched_class, pon_qos_add_sched_enter,
+	TP_PROTO(const struct pon_qos_qdisc *sch, u32 pid),
+	TP_ARGS(sch, pid)
+);
+
+DEFINE_EVENT(pon_qos_sched_class, pon_qos_add_sched_exit,
+	TP_PROTO(const struct pon_qos_qdisc *sch, u32 pid),
+	TP_ARGS(sch, pid)
+);
+
+DEFINE_EVENT(pon_qos_sched_class, pon_qos_sched_del_enter,
+	TP_PROTO(const struct pon_qos_qdisc *sch, u32 pid),
+	TP_ARGS(sch, pid)
+);
+
+DEFINE_EVENT(pon_qos_sched_class, pon_qos_sched_del_exit,
+	TP_PROTO(const struct pon_qos_qdisc *sch, u32 pid),
+	TP_ARGS(sch, pid)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_queue_class,
+	TP_PROTO(const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 int idx),
+
+	TP_ARGS(sch, q, idx),
+
+	TP_STRUCT__entry(
+		__string(name, sch->dev->name)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(u32, sch_id)
+		__field(u32, qid)
+		__field(int, arbi)
+		__field(int, p_w)
+		__field(int, idx)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, sch->dev->name);
+		__entry->handle = sch->handle;
+		__entry->parent = sch->handle;
+		__entry->sch_id = sch->sch_id;
+		__entry->qid = q->qid;
+		__entry->arbi = q->arbi;
+		__entry->p_w = q->p_w;
+		__entry->idx = idx;
+	),
+
+	TP_printk("[%s] hdls:%#x/%#x sch_id:%u qid:%u arbi:%i p_w:%i idx: %i",
+		  __get_str(name), __entry->parent, __entry->handle,
+		  __entry->sch_id, __entry->qid, __entry->arbi,
+		  __entry->p_w, __entry->idx)
+);
+
+DEFINE_EVENT(pon_qos_queue_class, pon_qos_queue_add_enter,
+	TP_PROTO(const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 int idx),
+	TP_ARGS(sch, q, idx)
+);
+
+DEFINE_EVENT(pon_qos_queue_class, pon_qos_queue_add_exit,
+	TP_PROTO(const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 int idx),
+	TP_ARGS(sch, q, idx)
+);
+
+DEFINE_EVENT(pon_qos_queue_class, pon_qos_queue_del_enter,
+	TP_PROTO(const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 int idx),
+	TP_ARGS(sch, q, idx)
+);
+
+DEFINE_EVENT(pon_qos_queue_class, pon_qos_queue_del_exit,
+	TP_PROTO(const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 int idx),
+	TP_ARGS(sch, q, idx)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_update_qmap_class,
+	TP_PROTO(const struct net_device *dev,
+		 const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 const struct pon_qos_qmap_tc *tc,
+		 bool en),
+
+	TP_ARGS(dev, sch, q, tc, en),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(s32, port)
+		__field(int, epn)
+		__field(u8, ds)
+		__field(u8, en)
+		__field(u32, qid)
+		__field(u32, classid)
+		__field(int, tc)
+		__field(char, tc_cookie)
+		__string(indev, tc->indev->name)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->port = sch->port;
+		__entry->epn = sch->epn;
+		__entry->ds = sch->ds;
+		__entry->en = en;
+		__entry->qid = q->qid;
+		__entry->classid = tc->handle;
+		__entry->tc = tc->tc;
+		__entry->tc_cookie = tc->tc_cookie;
+		__assign_str(indev, tc->indev ? tc->indev->name : "x");
+	),
+
+	TP_printk("[%s] en:%u ds:%u port:%i epn:%i q:%u cls:%#x tc:%i tc_cookie:%#x indev:%s",
+		  __get_str(name), __entry->en, __entry->ds,
+		  __entry->port, __entry->epn, __entry->qid,
+		  __entry->classid, __entry->tc, __entry->tc_cookie,
+		  __get_str(indev))
+);
+
+DEFINE_EVENT(pon_qos_update_qmap_class, pon_qos_update_qmap_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 const struct pon_qos_qmap_tc *tc,
+		 bool en),
+	TP_ARGS(dev, sch, q, tc, en)
+);
+
+DEFINE_EVENT(pon_qos_update_qmap_class, pon_qos_update_qmap_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 const struct pon_qos_qmap_tc *tc,
+		 bool en),
+	TP_ARGS(dev, sch, q, tc, en)
+);
+
+TRACE_EVENT(pon_qos_tc_red_replace,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_red_qopt_offload *opt,
+		 const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q,
+		 int type),
+
+	TP_ARGS(dev, opt, sch, q, type),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(u32, port)
+		__field(u32, sch_id)
+		__field(u32, qid)
+		__field(int, type)
+		__field(u32, min)
+		__field(u32, max)
+		__field(u64, probability)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->handle = opt->handle;
+		__entry->parent = opt->parent;
+		__entry->port = sch->port;
+		__entry->sch_id = sch->sch_id;
+		__entry->qid = q->qid;
+		__entry->type = type;
+		__entry->min = opt->set.min;
+		__entry->max = opt->set.max;
+		__entry->probability = opt->set.probability * 100;
+	),
+
+	TP_printk("[%s] hdls:%#x/%#x port:%u sch_id:%u qid:%u col:%s min:%u max:%u probability:%llu",
+		  __get_str(name), __entry->parent, __entry->handle,
+		  __entry->port, __entry->sch_id, __entry->qid,
+		  __entry->type == PON_QOS_QDATA_GREEN ? "green" : "yellow",
+		  __entry->min, __entry->max,
+		  DIV_ROUND_UP(
+		  DIV_ROUND_UP(__entry->probability, 1 << 16), 1 << 16))
+);
+
+TRACE_EVENT(pon_qos_tc_red_destroy,
+	TP_PROTO(const struct net_device *dev,
+		 const struct tc_red_qopt_offload *opt,
+		 const struct pon_qos_qdisc *sch,
+		 const struct pon_qos_q_data *q),
+
+	TP_ARGS(dev, opt, sch, q),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(u32, handle)
+		__field(u32, parent)
+		__field(u32, port)
+		__field(u32, sch_id)
+		__field(u32, qid)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->handle = opt->handle;
+		__entry->parent = opt->parent;
+		__entry->port = sch->port;
+		__entry->sch_id = sch->sch_id;
+		__entry->qid = q->qid;
+	),
+
+	TP_printk("[%s] hdls:%#x/%#x port:%u sch_id:%u qid:%u",
+		  __get_str(name), __entry->parent, __entry->handle,
+		  __entry->port, __entry->sch_id, __entry->qid)
+);
+
+DECLARE_EVENT_CLASS(pon_qos_tc_ext_vlan_class,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+
+	TP_ARGS(dev, ptr, tags, cnt, ingress),
+
+	TP_STRUCT__entry(
+		__string(name, dev->name)
+		__field(const void *, ptr)
+		__field(bool, ingress)
+		__field(int, tags)
+		__field(int, cnt)
+	),
+
+	TP_fast_assign(
+		__assign_str(name, dev->name);
+		__entry->ptr = ptr;
+		__entry->ingress = ingress;
+		__entry->tags = tags;
+		__entry->cnt = cnt;
+	),
+
+	TP_printk("[%s] -> node: %p %s tags:%i rules:%i",
+		  __get_str(name), __entry->ptr,
+		  __entry->ingress ? "ingress" : "egress",
+		  __entry->tags, __entry->cnt)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, ev_add_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, ev_add_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, ev_del_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, ev_del_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, vf_add_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, vf_add_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, vf_del_enter,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+DEFINE_EVENT(pon_qos_tc_ext_vlan_class, vf_del_exit,
+	TP_PROTO(const struct net_device *dev,
+		 const void *ptr, int tags, int cnt,
+		 bool ingress),
+	TP_ARGS(dev, ptr, tags, cnt, ingress)
+);
+
+#endif /* __PON_QOS_DEVICE_TRACE_DATA */
+
+#undef TRACE_INCLUDE_PATH
+#define TRACE_INCLUDE_PATH .
+#undef TRACE_INCLUDE_FILE
+#define TRACE_INCLUDE_FILE pon_qos_trace
+
+#include <trace/define_trace.h>
